src/threadLocalData.h

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _ASPROF_THREAD_LOCAL_H
+#define _ASPROF_THREAD_LOCAL_H
+
+#include "asprof.h"
+#include "mutex.h"
+#include "arguments.h"
+#include <pthread.h>
+#include <atomic>
+
+class ThreadLocalData {
+public:
+    // increment the thread-local sample counter. See the `asprof_thread_local_data` docs.
+    //
+    // this function *is* async-signal safe, and therefore will not initialize the thread-local
+    // storage by itself, but will rather do nothing if it's not initialized already. This works
+    // fine with the `sample_counter` API since only changes in `sample_counter` matter.
+    static void incrementSampleCounter(void)  {
+        if (_profiler_data_key == -1) return;
+
+        asprof_thread_local_data *data = (asprof_thread_local_data *)pthread_getspecific(_profiler_data_key);
+        if (data != NULL) {
+            data->sample_counter++;
+        }
+    }
+
+    // Get the `asprof_thread_local_data`. See the `asprof_get_thread_local_data` docs.
+    static asprof_thread_local_data *getThreadLocalData(void)  {
+        if (_profiler_data_key == -1) {
+            return NULL;
+        }
+
+        asprof_thread_local_data *val = (asprof_thread_local_data*) pthread_getspecific(_profiler_data_key);
+        if (val == NULL) {
+            return initPerThreadThreadLocalData(_profiler_data_key);
+        }
+
+        return val;
+    }
+private:
+    static asprof_thread_local_data *initPerThreadThreadLocalData(pthread_key_t profiler_data_key);
```

Comment:
"PerThreadThreadLocal" is a kind of tautology.
Also, the function does not need to accept an argument, since it's always `_profiler_data_key` accessible from that function.